enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,441 @@
+providers are similar, but not identical.
+
+This enhancement will describe how we can ensure correct configuration is passed
+to the cloud provider in use during and after an upgrade, and how we can prevent
+an upgrade to 4.11 if the configuration cannot be upgraded.
+
+## Motivation
+
+The legacy cloud provider is dead. Long live the external cloud provider.
+
+### Goals
+
+- No user action is required in the common case
+- Upgrade legacy cloud provider to external cloud controller manager where possible
+- Mark cluster as non-upgradable where upgrade to external CCM is not possible
+
+### Non-Goals
+
+- Modify user-managed configuration
+- Manage cloud providers for other, non-OpenStack platforms
+
+## Proposal
+
+- The change will be implemented in the Cluster Cloud Controller Manager
+  Operator (CCCMO)
+- CCCMO will create a separate configmap for the external cloud provider
+- We will prevent upgrade if it is not possible to create the external cloud
+  provider configmap
+
+### Implemented in the Cluster Cloud Controller Manager Operator (CCCMO)
+
+The [Cluster Config Operator (CCO)][CCO] is currently responsible for managing
+the config for the legacy cloud provider. This applies a transformation function
+to the user-managed config and writes the result to
+`openshift-config-managed/kube-cloud-config`. The transformation
+function is selected according to PlatformType in the cluster Infrastructure
+object. Components requiring legacy cloud provider configuration, including
+kubelet and KCM, read the transformed configmap.
+
+CCO will **not** be modified. The new functionality will instead be implemented
+as a change to the [Cluster Cloud Controller Manager Operator (CCCMO)][CCCMO],
+functioning in much the same way as CCO's management of the legacy
+configuration.
+
+[CCO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-config-operator_platform-operators-ref
+[CCCMO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-cloud-controller-manager-operator_platform-operators-ref
+
+### Separate ConfigMaps for legacy and external cloud providers
+
+The cloud configuration is read by multiple components which will not all be
+upgraded simultaneously. If we simply updated the
+`openshift-config-managed/kube-cloud-config` to be compatible with external CCM
+we would risk breaking KCM and kubelets which have not yet been upgraded.
+Therefore we are constrained to write two versions of the managed config map.
+
+CCO will be continue to create `openshift-config-managed/kube-cloud-config`,
+which will remain compatible with the legacy cloud provider. CCCMO will create a
+new `openshift-config-managed/cloud-controller-manager-config` config map, which
+will be compatible with the external CCM. The external CCM will then read its
+configuration from `openshift-config-managed/cloud-controller-manager-config`.
```

Comment:
Why is it not reading it from the namespace in which it exists? Do we need to put it in the managed namespace?

Reply:
+1.

Reply:
wdym? We don't read from `openshift-config/cloud-provider-config` because that's intended to be compatible with both legacy and external cloud providers for now, and even after the legacy cloud provider goes there will be config that we won't want the user to be able to set. Is there yet another namespace we should be looking at?

Reply:
That's a good point. If we're no longer using CCO there's no reason we can't write the transformed config to `openshift-cloud-controller-manager`, which is where OCCM is running.

Reply:
Yup, that's what I mean, that way config is easily discoverable.

Reply:
Yep, `openshift-cloud-controller-manager` is what I was referring to, we should write the configmap directly to this namespace as it is the only namespace in which it will actually be used. If we write it to `openshift-config`, something then needs to copy it to `openshift-cloud-controller-manager` else the pod won't be able to mount it

Reply:
Updated to use `openshift-cloud-controller-manager/cloud-conf`.